File: backend/functions/chatbot.py
import json
import os
from datetime import datetime, timedelta
from dateparser.search import search_dates
from openai import OpenAI
import requests

# Read Environment Vars (Netlify dashboard)
OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
HOTEL_API_ENDPOINT = "https://hotel.dev-maister.gr/hotel_Casa/mcp_server/index.php"
HOTEL_API_BEARER_TOKEN = os.getenv("HOTEL_API_BEARER_TOKEN")

client = OpenAI(api_key=OPENAI_API_KEY)

# -------------------------
#  HOTEL AVAILABILITY
# -------------------------
def get_hotel_availability(json_key: str, start: str, end: str, adults: str, kids: str, minors: str):
    headers = {
        "Authorization": f"Bearer {HOTEL_API_BEARER_TOKEN}",
        "Accept": "*/*",
        "User-Agent": "Mozilla/5.0"
    }
    params = {
        "json_key": json_key,
        "start": start,
        "end": end,
        "adults": adults,
        "kids": kids,
        "minors": minors
    }
    try:
        response = requests.get(HOTEL_API_ENDPOINT, headers=headers, params=params, timeout=10)
        response.raise_for_status()
        return response.json()
    except Exception as e:
        return {"error": "API call failed", "details": str(e)}

# -------------------------
#  HOTEL PRICE
# -------------------------
def get_hotel_price(json_key: str, start: str, end: str, adults: str, kids: str, minors: str):
    headers = {
        "Authorization": f"Bearer {HOTEL_API_BEARER_TOKEN}",
        "Accept": "*/*",
        "User-Agent": "Mozilla/5.0"
    }
    params = {
        "json_key": json_key,
        "start": start,
        "end": end,
        "adults": adults,
        "kids": kids,
        "minors": minors
    }
    try:
        response = requests.get(HOTEL_API_ENDPOINT, headers=headers, params=params, timeout=10)
        response.raise_for_status()
        return response.json()
    except Exception as e:
        return {"error": "API call failed", "details": str(e)}

# -------------------------
#  TOOLS
# -------------------------
tool_availability = {
    "name": "get_hotel_availability",
    "description": "Get hotel room availability",
    "parameters": {
        "type": "object",
        "properties": {
            "json_key": {"type": "string"},
            "start": {"type": "string"},
            "end": {"type": "string"},
            "adults": {"type": "string"},
            "kids": {"type": "string"},
            "minors": {"type": "string"}
        },
        "required": ["json_key", "start", "end", "adults", "kids", "minors"]
    }
}

tool_price = {
    "name": "get_hotel_price",
    "description": "Get hotel room prices",
    "parameters": {
        "type": "object",
        "properties": {
            "json_key": {"type": "string"},
            "start": {"type": "string"},
            "end": {"type": "string"},
            "adults": {"type": "string"},
            "kids": {"type": "string"},
            "minors": {"type": "string"}
        },
        "required": ["json_key", "start", "end", "adults", "kids", "minors"]
    }
}

system_prompt = """
You are a hotel reception assistant for Hotel Ilion.
You can chat naturally...
(unchanged)
"""

# -------------------------
#  GPT WRAPPER
# -------------------------
def call_gpt(messages, tools=None):
    resp = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=messages,
        functions=(tools if tools else []),
        function_call="auto"
    )
    return resp.choices[0].message

# -------------------------
#  PROCESS TOOL CALL
# -------------------------
def try_handle_tool_call(ai_message):
    fc = ai_message.function_call
    if not fc:
        return None

    args = json.loads(fc.arguments or "{}")

    if fc.name == "get_hotel_availability":
        result = get_hotel_availability(**args)
        return {"role": "function", "name": fc.name, "content": json.dumps(result)}

    if fc.name == "get_hotel_price":
        result = get_hotel_price(**args)
        return {"role": "function", "name": fc.name, "content": json.dumps(result)}

    return None

# =================================================
#  NETLIFY FUNCTION HTTP HANDLER
# =================================================
def handler(event, context):

    try:
        body = json.loads(event.get("body", "{}"))
        user_message = body.get("message", "")
        conversation = body.get("messages", [])

        if not conversation:
            conversation = [{"role": "system", "content": system_prompt}]

        conversation.append({"role": "user", "content": user_message})

        ai_msg = call_gpt(conversation, tools=[tool_availability, tool_price])
        conversation.append({
            "role": "assistant",
            "content": ai_msg.content,
            "function_call": ai_msg.function_call
        })

        tool_response = try_handle_tool_call(ai_msg)

        if tool_response:
            conversation.append(tool_response)
            ai_msg2 = call_gpt(conversation, tools=[tool_availability, tool_price])
            return {
                "statusCode": 200,
                "body": json.dumps({
                    "response": ai_msg2.content,
                    "messages": conversation
                })
            }

        return {
            "statusCode": 200,
            "body": json.dumps({
                "response": ai_msg.content,
                "messages": conversation
            })
        }

    except Exception as e:
        return {
            "statusCode": 500,
            "body": json.dumps({"error": str(e)})
        }

File: functions/chatbot.js
// chatbot.js - Netlify Function

const fetch = require("node-fetch"); // Node-fetch v2
const OpenAI = require("openai");

// --- OpenAI client ---
const client = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY
});

// --- Hotel API ---
const HOTEL_API_ENDPOINT = "https://hotel.dev-maister.gr/hotel_Casa/mcp_server/index.php";
const HOTEL_API_BEARER_TOKEN = process.env.HOTEL_API_BEARER_TOKEN;

// --- Helper functions ---
async function getHotelAvailability({ json_key, start, end, adults, kids, minors }) {
  const headers = {
    Authorization: `Bearer ${HOTEL_API_BEARER_TOKEN}`,
    Accept: "*/*",
    "User-Agent": "Node.js"
  };
  const params = new URLSearchParams({ json_key, start, end, adults, kids, minors });
  try {
    const res = await fetch(`${HOTEL_API_ENDPOINT}?${params.toString()}`, { headers });
    return await res.json();
  } catch (err) {
    return { error: "API call failed", details: err.message };
  }
}

async function getHotelPrice({ json_key, start, end, adults, kids, minors }) {
  const headers = {
    Authorization: `Bearer ${HOTEL_API_BEARER_TOKEN}`,
    Accept: "*/*",
    "User-Agent": "Node.js"
  };
  const params = new URLSearchParams({ json_key, start, end, adults, kids, minors });
  try {
    const res = await fetch(`${HOTEL_API_ENDPOINT}?${params.toString()}`, { headers });
    return await res.json();
  } catch (err) {
    return { error: "API call failed", details: err.message };
  }
}

// --- Netlify handler ---
exports.handler = async function(event, context) {
  // --- CORS preflight ---
  if (event.httpMethod === "OPTIONS") {
    return {
      statusCode: 204,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Headers": "Content-Type",
        "Access-Control-Allow-Methods": "POST, OPTIONS"
      },
      body: ""
    };
  }

  try {
    const body = JSON.parse(event.body || "{}");
    const userMessage = body.message || "";
    const conversation = body.messages || [];

    const systemPrompt = `
      You are a hotel reception assistant for Hotel Ilion.
      You can chat naturally with the user, ask clarifying questions, and only call 
      getHotelAvailability or getHotelPrice when you have check-in, check-out, and guests.
      Never reveal room names or numbers.
      Always clarify ages of children.
    `;

    const messages = [
      { role: "system", content: systemPrompt },
      ...conversation,
      { role: "user", content: userMessage }
    ];

    const tools = [
      {
        name: "getHotelAvailability",
        description: "Get hotel room availability for a given date range.",
        parameters: {
          type: "object",
          properties: {
            json_key: { type: "string" },
            start: { type: "string" },
            end: { type: "string" },
            adults: { type: "string" },
            kids: { type: "string" },
            minors: { type: "string" }
          },
          required: ["json_key", "start", "end", "adults", "kids", "minors"]
        }
      },
      {
        name: "getHotelPrice",
        description: "Get hotel room prices for a given date range.",
        parameters: {
          type: "object",
          properties: {
            json_key: { type: "string" },
            start: { type: "string" },
            end: { type: "string" },
            adults: { type: "string" },
            kids: { type: "string" },
            minors: { type: "string" }
          },
          required: ["json_key", "start", "end", "adults", "kids", "minors"]
        }
      }
    ];

    // --- GPT Call ---
    const gptResponse = await client.chat.completions.create({
      model: "gpt-4o-mini",
      messages,
      functions: tools,
      function_call: "auto"
    });

    let aiMsg = gptResponse.choices[0].message;

    if (aiMsg.function_call) {
      const { name, arguments: argsStr } = aiMsg.function_call;
      const args = JSON.parse(argsStr || "{}");
      let result;

      if (name === "getHotelAvailability") result = await getHotelAvailability(args);
      else if (name === "getHotelPrice") result = await getHotelPrice(args);
      else result = { error: "Unknown function call" };

      messages.push({ role: "assistant", content: null, function_call: aiMsg.function_call });
      messages.push({ role: "function", name, content: JSON.stringify(result) });

      const finalGpt = await client.chat.completions.create({
        model: "gpt-4o-mini",
        messages,
        functions: tools
      });

      aiMsg = finalGpt.choices[0].message;
    }

    // --- Success response with CORS ---
    return {
      statusCode: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Headers": "Content-Type"
      },
      body: JSON.stringify({ response: aiMsg.content || "Hello!", messages })
    };

  } catch (err) {
    return {
      statusCode: 500,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Headers": "Content-Type"
      },
      body: JSON.stringify({ error: err.message })
    };
  }
};
